engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java

```diff
@@ -279,25 +274,36 @@ private void updateAvailableFeatures() {
         }
     }
 
-    private void compileShaders(Set<ShaderProgramFeature> features) {
-        compileShader(GL20.GL_FRAGMENT_SHADER, features);
-        compileShader(GL20.GL_VERTEX_SHADER, features);
-    }
-
     private void compileAllShaderPermutations() {
-        compileShaders(Collections.<ShaderProgramFeature>emptySet());
+        int fragShaderId = compileShader(GL20.GL_FRAGMENT_SHADER, Collections.emptySet());
+        int vertShaderId = compileShader(GL20.GL_VERTEX_SHADER, Collections.emptySet());
 
-        int counter = 1;
+        if (!compileSuccess(fragShaderId) || !compileSuccess(vertShaderId)) {
 
-        for (Set<ShaderProgramFeature> permutation : Sets.powerSet(availableFeatures)) {
-            compileShaders(permutation);
-            counter++;
+            String errorMessage = String.format("Shader '%s' failed to compile.\n\n"
+                    + "Vertex Shader Info: \n%s\n"
+                    + "Fragment Shader Info: \n%s",
+                    getURI(), getLogInfo(vertShaderId), getLogInfo(fragShaderId));
+
+            throw new RuntimeException(errorMessage);
         }
-        logger.debug("Compiled {} permutations for {}.", counter, getURI());
+
+        Set<Set<ShaderProgramFeature>> allPermutations = Sets.powerSet(availableFeatures);
+
+        for (Set<ShaderProgramFeature> permutation : allPermutations) {
+            fragShaderId = compileShader(GL20.GL_FRAGMENT_SHADER, permutation);
+            vertShaderId = compileShader(GL20.GL_VERTEX_SHADER, permutation);
+
+            if (!compileSuccess(fragShaderId) || !compileSuccess(vertShaderId)) {
+                throw new RuntimeException(String.format("Shader '%s' failed to compile for feature set '%s'",
+                        getURI(), permutation));
+            }
+        }
+
+        logger.debug("Compiled {} permutations for {}.", allPermutations.size(), getURI());
     }
 
-    private void compileShader(int type, Set<ShaderProgramFeature> features) {
-        int shaderId = GL20.glCreateShader(type);
+    private String createShader(int type, Set<ShaderProgramFeature> features) {
```

Comment:
I'm thinking this could be better called buildShader or assembleShader, also to differentiate it more from glCreateShader.